# **SM Heart Disease Kaggle Project**
Part 2: Modeling

I created this Modeling notebook in continuation of my EDA notebook on the heart disease dataset found here: https://www.kaggle.com/ronitf/heart-disease-uci

***

**Libraries and overview of original dataset**


## **Load libraries**

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import numpy as np
from sklearn.linear_model import LogisticRegression 
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier 
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification

## **Load dataset**

In [2]:
#load the data
#df = pd.read_csv('heart.csv')
#from google.colab import files
#uploaded = files.upload()

ModuleNotFoundError: No module named 'google.colab'

In [4]:
df = pd.read_csv('heart.csv')

**Recap of important data set info from Part 1 EDA Notebook**

 df.head gives us the top 5 entries and is common for df exploration 

In [5]:
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

df.info gives us an overview and is common for df exploration 

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
age         303 non-null int64
sex         303 non-null int64
cp          303 non-null int64
trestbps    303 non-null int64
chol        303 non-null int64
fbs         303 non-null int64
restecg     303 non-null int64
thalach     303 non-null int64
exang       303 non-null int64
oldpeak     303 non-null float64
slope       303 non-null int64
ca          303 non-null int64
thal        303 non-null int64
target      303 non-null int64
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


df.describe gives summary statistics of variables and is common for df exploration 

In [7]:
df.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

**Predictive Modeling, Whole Data Set**

Train_Test_Split:

We will use "from sklearn.model_selection import train_test_split" to separate our data into a training group and test group for our models.

In [8]:
df['target'] = df['target'].replace(to_replace = [0, 1], value = [True, False])

In [9]:
X = df.drop(columns=['target'])
y= df['target']

In [10]:
r_state = 35

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=r_state) 


In [12]:
X_train.shape

(203, 13)

In [13]:
X_test.shape

(100, 13)

**LogisticRegression Whole**




We will use "from sklearn.linear_model import LogisticRegression" to form and assess a Logistic Regression model.

In [14]:
my_model = LogisticRegression(random_state=r_state)
my_model.fit(X_train, y_train)

C:\Users\S\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=35, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [15]:
y_pred = my_model.predict(X_test)

In [16]:
y_pred

array([False,  True,  True,  True, False, False,  True, False, False,
       False,  True, False,  True, False,  True, False, False, False,
       False,  True, False, False, False,  True, False,  True,  True,
       False, False, False, False, False, False,  True, False, False,
       False,  True,  True,  True,  True,  True, False,  True,  True,
        True,  True, False, False,  True, False, False,  True,  True,
       False,  True, False, False,  True, False,  True, False,  True,
        True, False,  True, False,  True, False,  True, False,  True,
        True,  True,  True,  True,  True, False, False,  True, False,
       False,  True, False, False,  True, False,  True,  True,  True,
        True, False, False,  True,  True, False, False,  True, False,
       False])

In [17]:
y_test.to_numpy()

array([ True,  True,  True,  True, False, False,  True, False, False,
       False, False, False,  True, False, False, False, False, False,
       False, False, False, False, False,  True, False,  True,  True,
       False, False, False, False, False, False,  True, False, False,
        True,  True,  True,  True,  True,  True, False,  True,  True,
        True,  True,  True, False,  True, False,  True, False, False,
       False, False, False,  True,  True, False,  True, False,  True,
        True,  True,  True, False,  True, False,  True, False, False,
        True,  True, False,  True,  True, False,  True, False, False,
       False, False, False, False,  True, False,  True,  True,  True,
        True, False, False,  True,  True, False, False,  True,  True,
       False])

In [18]:
my_model.score(X_test, y_test)

0.82

For this my_model, in Colab, the multi_class is set to auto and the solver is set to lbfgs; in Jupyter Notebook, multi_class is set to warn and solver is set to warn. This may explain a .02 higher my_model.score of .84 above (in Colab).

**Sub models based on correlated features, applied to Whole Data Set:**

These features show the greatest negative correlations to the target: exang, oldpeak, ca, and thal

This submodel scores lower than the whole data set with all features.

In [19]:
columns = ['exang', 'oldpeak', 'ca', 'thal']
subset = df[columns]
subset.head()

exang  oldpeak  ca  thal
0      0      2.3   0     1
1      0      3.5   0     2
2      0      1.4   0     2
3      0      0.8   0     2
4      1      0.6   0     2

In [20]:
X = subset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=r_state) 


In [21]:
sub_model = LogisticRegression(random_state=r_state)
sub_model.fit(X_train, y_train)
y_pred =sub_model.predict(X_test)
sub_model.score(X_test, y_test)

C:\Users\S\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.78

These features show the greatest positive correlations to the target: cp, thalach, and slope 

This submodel scores lower than the whole data set with all features, as well as the submodel above based on greatest negative correlations to target.

In [22]:
columns = ['cp', 'slope', 'thalach']
subset = df[columns]
subset.head()

cp  slope  thalach
0   3      0      150
1   2      0      187
2   1      2      172
3   1      2      178
4   0      2      163

In [23]:
X = subset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=r_state) 


In [24]:
sub_model = LogisticRegression(random_state=r_state)
sub_model.fit(X_train, y_train)
y_pred =sub_model.predict(X_test)
sub_model.score(X_test, y_test)

C:\Users\S\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.78

##**One-hot encoded dataset for features sex, cp, slope and thal:**

In [25]:
#df.head will show columns for original dataframe
dataset = pd.get_dummies(df, columns = ['sex','cp', 'slope', 'thal'])
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1   False  
1   0     2   False  
2   0     2   False  
3   0     2   False  
4   0     2   False

In [26]:
#dataset.head will show columns for one-hot encoded sex, cp, slope, and thal features:
dataset.head()

age  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  ca  target  \
0   63       145   233    1        0      150      0      2.3   0   False   
1   37       130   250    0        1      187      0      3.5   0   False   
2   41       130   204    0        0      172      0      1.4   0   False   
3   56       120   236    0        1      178      0      0.8   0   False   
4   57       120   354    0        1      163      1      0.6   0   False   

   ...  cp_1  cp_2  cp_3  slope_0  slope_1  slope_2  thal_0  thal_1  thal_2  \
0  ...     0     0     1        1        0        0       0       1       0   
1  ...     0     1     0        1        0        0       0       0       1   
2  ...     1     0     0        0        0        1       0       0       1   
3  ...     1     0     0        0        0        1       0       0       1   
4  ...     0     0     0        0        0        1       0       0       1   

   thal_3  
0       0  
1       0  
2       0  
3       0  
4       0  

[5 rows x 23 columns]

In [27]:
dX = dataset.drop(columns=['target'])
dy= dataset['target']

In [28]:
r_state = 35

In [29]:
dX_train, dX_test, dy_train, dy_test = train_test_split(dX, dy, test_size=0.33, random_state=r_state) 


In [30]:
dX_train.shape

(203, 22)

In [31]:
dX_test.shape

(100, 22)

In [32]:
d_model = LogisticRegression(random_state=r_state)
d_model.fit(dX_train, dy_train)

C:\Users\S\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=35, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [33]:
dy_pred = d_model.predict(dX_test)

In [34]:
dy_pred

array([False,  True,  True,  True, False, False,  True, False, False,
       False,  True, False,  True, False,  True, False, False, False,
       False, False, False, False, False,  True, False,  True,  True,
       False, False, False, False, False, False,  True, False, False,
        True,  True,  True, False,  True,  True, False,  True,  True,
        True,  True, False, False,  True, False, False, False,  True,
       False,  True, False, False,  True, False,  True, False,  True,
        True, False,  True, False,  True,  True,  True, False,  True,
        True,  True,  True,  True,  True, False, False, False, False,
       False,  True, False, False,  True, False,  True,  True,  True,
        True, False, False,  True,  True, False, False,  True, False,
       False])

In [35]:
dy_test.to_numpy()

array([ True,  True,  True,  True, False, False,  True, False, False,
       False, False, False,  True, False, False, False, False, False,
       False, False, False, False, False,  True, False,  True,  True,
       False, False, False, False, False, False,  True, False, False,
        True,  True,  True,  True,  True,  True, False,  True,  True,
        True,  True,  True, False,  True, False,  True, False, False,
       False, False, False,  True,  True, False,  True, False,  True,
        True,  True,  True, False,  True, False,  True, False, False,
        True,  True, False,  True,  True, False,  True, False, False,
       False, False, False, False,  True, False,  True,  True,  True,
        True, False, False,  True,  True, False, False,  True,  True,
       False])

In [36]:
d_model.score(dX_test, dy_test)

0.84

**One-hot encoding seemed to help raise our model's score by 2%**




**RandomForestClassifier Whole**



Resetting dataset so target is not set to True/False from LogisticRegression above

In [37]:
df_Rand = pd.read_csv('heart.csv')

In [38]:
X = df_Rand.drop(columns=['target'])
y= df_Rand['target']

In [39]:
r_state = 35

In [40]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=r_state) 


Decision Tree for comparison


In [41]:
DT_whole = DecisionTreeClassifier(random_state = 35)
DT_whole.fit(X_train, y_train)
y_pred = DT_whole.predict(X_test)
accuracy_score(y_test, y_pred)

0.74

We will use "from sklearn.ensemble import RandomForestRegressor" to form and assess a Random Forests model.

In [42]:
RF = RandomForestClassifier()
RF.fit(X_train, y_train)
y_pred = RF.predict(X_test)
accuracy_score(y_test, y_pred)

C:\Users\S\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.8

In [43]:
Rand = RandomForestClassifier(n_estimators=100, random_state=35)
Rand.fit(X_train, y_train)
y_pred = Rand.predict(X_test)
accuracy_score(y_test, y_pred)


0.8

In [44]:
Rand_whole = RandomForestClassifier(n_estimators=10, random_state=35)
Rand_whole.fit(X_train, y_train)
y_pred = Rand_whole.predict(X_test)
accuracy_score(y_test, y_pred)



0.75

**Predictive Modeling, Female Data Set**

Making female only data set:

In [45]:
df_sex = pd.read_csv('heart.csv')


In [46]:
df_fem =  df_sex['sex']==0
print(df_fem.head())

0    False
1    False
2     True
3    False
4     True
Name: sex, dtype: bool


In [47]:
# filter rows for females only using  the boolean variable
fem_only = df_sex[df_fem]
print(fem_only.shape)

(96, 14)


In [48]:
print(fem_only.head())

    age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
2    41    0   1       130   204    0        0      172      0      1.4   
4    57    0   0       120   354    0        1      163      1      0.6   
6    56    0   1       140   294    0        0      153      0      1.3   
11   48    0   2       130   275    0        1      139      0      0.2   
14   58    0   3       150   283    1        0      162      0      1.0   

    slope  ca  thal  target  
2       2   0     2       1  
4       2   0     2       1  
6       1   0     2       1  
11      2   0     2       1  
14      2   0     2       1  


In [49]:
fem_only.describe()

age   sex         cp    trestbps        chol        fbs  \
count  96.000000  96.0  96.000000   96.000000   96.000000  96.000000   
mean   55.677083   0.0   1.041667  133.083333  261.302083   0.125000   
std     9.409396   0.0   0.972427   19.311119   65.088946   0.332455   
min    34.000000   0.0   0.000000   94.000000  141.000000   0.000000   
25%    49.750000   0.0   0.000000  120.000000  214.750000   0.000000   
50%    57.000000   0.0   1.000000  131.000000  253.000000   0.000000   
75%    63.000000   0.0   2.000000  140.000000  296.750000   0.000000   
max    76.000000   0.0   3.000000  200.000000  564.000000   1.000000   

         restecg     thalach      exang    oldpeak      slope         ca  \
count  96.000000   96.000000  96.000000  96.000000  96.000000  96.000000   
mean    0.572917  151.125000   0.229167   0.876042   1.427083   0.552083   
std     0.557150   20.047969   0.422503   1.119844   0.593736   0.881026   
min     0.000000   96.000000   0.000000   0.000000   0.000000   0.000000   
25%     0.000000  141.250000   0.000000   0.000000   1.000000   0.000000   
50%     1.000000  157.000000   0.000000   0.600000   1.000000   0.000000   
75%     1.000000  165.000000   0.000000   1.400000   2.000000   1.000000   
max     2.000000  192.000000   1.000000   6.200000   2.000000   3.000000   

           thal     target  
count  96.00000  96.000000  
mean    2.12500   0.750000  
std     0.44129   0.435286  
min     0.00000   0.000000  
25%     2.00000   0.750000  
50%     2.00000   1.000000  
75%     2.00000   1.000000  
max     3.00000   1.000000

__Train_Test_Split__

We will use "from sklearn.model_selection import train_test_split" to separate our data into a training group and test group for our models.

In [50]:
fX = fem_only.drop(columns=['target'])
fy= fem_only['target']

In [51]:
r_state = 35

In [52]:
fX_train, fX_test, fy_train, fy_test = train_test_split(fX, fy, test_size=0.33, random_state=r_state) 


In [53]:
fX_train.shape

(64, 13)

In [54]:
fX_test.shape

(32, 13)

**LogisticRegression Female Only**




We will use "from sklearn.linear_model import LogisticRegression" to form and assess a Logistic Regression model.

In [55]:
my_fmodel = LogisticRegression(random_state=r_state)
my_fmodel.fit(fX_train, fy_train)

C:\Users\S\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=35, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [56]:
fy_pred = my_fmodel.predict(fX_test)

In [57]:
fy_pred

array([1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 0, 0, 1, 1, 1], dtype=int64)

In [58]:
fy_test.to_numpy()

array([1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 0, 1, 1, 0], dtype=int64)

In [59]:
my_fmodel.score(fX_test, fy_test)

0.78125

**RandomForestClassifier Female Only**



Resetting dataset so target is not set to True/False from LogisticRegression above

Decision Tree for comparison


In [60]:
DT_fem = DecisionTreeClassifier(random_state = 35)
DT_fem.fit(fX_train, fy_train)
fy_pred = DT_fem.predict(fX_test)
accuracy_score(fy_test, fy_pred)

0.8125

We will use "from sklearn.ensemble import RandomForestClassifier" to form and assess a Random Forests model.

In [61]:
RF_fem = RandomForestClassifier()
RF_fem.fit(fX_train, fy_train)
fy_pred = RF_fem.predict(fX_test)
accuracy_score(fy_test, fy_pred)

C:\Users\S\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.78125

In [62]:
Rand_fem = RandomForestClassifier(n_estimators=100, random_state=35)
Rand_fem.fit(fX_train, fy_train)
fy_pred = Rand_fem.predict(fX_test)
accuracy_score(fy_test, fy_pred)


0.875

In [63]:
Rand_f = RandomForestClassifier(n_estimators=10, random_state=35)
Rand_f.fit(fX_train, fy_train)
fy_pred = Rand_f.predict(fX_test)
accuracy_score(fy_test, fy_pred)



0.78125

**Predictive Modeling, Male Data Set**

Making male only data set:

In [64]:
df_male =  df_sex['sex']==1
print(df_male.head())

0     True
1     True
2    False
3     True
4    False
Name: sex, dtype: bool


In [65]:
# filter rows for males only using  the boolean variable
male_only = df_sex[df_male]
print(male_only.shape)

(207, 14)


In [66]:
print(male_only.head())

   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
3   56    1   1       120   236    0        1      178      0      0.8      2   
5   57    1   0       140   192    0        1      148      0      0.4      1   
7   44    1   1       120   263    0        1      173      0      0.0      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
3   0     2       1  
5   0     1       1  
7   0     3       1  


In [67]:
male_only.describe()

age    sex          cp    trestbps        chol         fbs  \
count  207.000000  207.0  207.000000  207.000000  207.000000  207.000000   
mean    53.758454    1.0    0.932367  130.946860  239.289855    0.159420   
std      8.883803    0.0    1.059064   16.658246   42.782392    0.366955   
min     29.000000    1.0    0.000000   94.000000  126.000000    0.000000   
25%     47.000000    1.0    0.000000  120.000000  208.000000    0.000000   
50%     54.000000    1.0    0.000000  130.000000  235.000000    0.000000   
75%     59.500000    1.0    2.000000  140.000000  268.000000    0.000000   
max     77.000000    1.0    3.000000  192.000000  353.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  207.000000  207.000000  207.000000  207.000000  207.000000  207.000000   
mean     0.507246  148.961353    0.371981    1.115459    1.386473    0.811594   
std      0.510754   24.130882    0.484505    1.174632    0.627378    1.074082   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  132.000000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  151.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  168.000000    1.000000    1.800000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    5.600000    2.000000    4.000000   

             thal      target  
count  207.000000  207.000000  
mean     2.400966    0.449275  
std      0.659949    0.498626  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    0.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

__Train_Test_Split__

We will use "from sklearn.model_selection import train_test_split" to separate our data into a training group and test group for our models.

In [68]:
mX = male_only.drop(columns=['target'])
my= male_only['target']

In [69]:
r_state = 35

In [70]:
mX_train, mX_test, my_train, my_test = train_test_split(mX, my, test_size=0.33, random_state=r_state) 


In [71]:
mX_train.shape

(138, 13)

In [72]:
mX_test.shape

(69, 13)

**LogisticRegression Male Only**




We will use "from sklearn.linear_model import LogisticRegression" to form and assess a Logistic Regression model.

In [73]:
mmodel = LogisticRegression(random_state=r_state)
mmodel.fit(mX_train, my_train)

C:\Users\S\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=35, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [74]:
my_pred = mmodel.predict(mX_test)

In [75]:
my_pred

array([1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 1, 1], dtype=int64)

In [76]:
my_test.to_numpy()

array([1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 1], dtype=int64)

In [77]:
mmodel.score(mX_test, my_test)

0.7681159420289855

**RandomForestClassifier Male Only**



Decision Tree for comparison


In [78]:
DT_male = DecisionTreeClassifier(random_state = 35)
DT_male.fit(mX_train, my_train)
my_pred = DT_male.predict(mX_test)
accuracy_score(my_test, my_pred)

0.6086956521739131

We will use "from sklearn.ensemble import RandomForestRegressor" to form and assess a Random Forests model.

In [79]:
RF_male = RandomForestClassifier()
RF_male.fit(mX_train, my_train)
my_pred = RF_male.predict(mX_test)
accuracy_score(my_test, my_pred)

C:\Users\S\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.7246376811594203

In [80]:
Rand_male = RandomForestClassifier(n_estimators=100, random_state=35)
Rand_male.fit(mX_train, my_train)
my_pred = Rand_male.predict(mX_test)
accuracy_score(my_test, my_pred)


0.6811594202898551

In [81]:
Rand_m = RandomForestClassifier(n_estimators=10, random_state=35)
Rand_m.fit(mX_train, my_train)
fy_pred = Rand_m.predict(mX_test)
accuracy_score(my_test, my_pred)



0.6811594202898551

# **Ending Thoughts and Action Steps**
(Conclusion references model outcomes from Google Colab notebook, in case any change in numbers here in Jupyter)


**The best model was the Random Forest for the Female Only data set**, with Random Forest no parameters and Random Forest (n_estimators=100, random_state=35) tying with a score of .875. 

Coming in after that was the Logistic Regression of the Whole data set with One Hot Encoded categorical features with a score of .86.

**The worst model was the simple Decision Tree for the Male Only dataset** with a score of: 0.6086956521739131. The Logistic Regression model was able to bump it up to 0.7681159420289855.

---

This project was really interesting. **A surprising curveball is how the target--what this entire project focuses on--led to much confusion online as to what these values mean.** At first it seemed that 1 = heart disease, but upon multiple Kaggle references, it seems that 0 = heart disease. 

The original source should be more clear on what the features mean and how they should be interpreted. I think the confusion could just be in how the < and > signs are written. When you read -- Value 0: < 50% diameter narrowing, it sounds like less than half of the blood vessel is narrow, which is healthy. I now think that the author is using it more as an equation, so that "50% diameter narrowing" is the greater amount. Vice versa for > 50% diameter narrowing.
https://archive.ics.uci.edu/ml/datasets/Heart+Disease

---

Another large point of interest is how features related to female health vs. male health. It is well known now that there are many differences based on sex in how cardiovascular disease is recognized and diagnosed.

"Cardiovascular disease is the major cause of death in women and is still under-recognised and undertreated. A greater awareness of the differences in presentation of angina pectoris and ACS between men and women, with gender-based interpretation of diagnostic tests, is mandatory for health care professionals to improve therapeutic strategies and outcomes in women. Cardiology guidelines should be more focused on sex-related differences when appropriate. Further, women themselves need to be more aware of their own risk factors and clinical signs of CHD. Many biological differences in atherosclerosis between men and women are not yet clarified and will need further research in the future."
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3018605/



### **Similar to learned biases in machine learning models and artificial intelligence applications, more awareness of biased medical models is now leading toward better treatment of many ailments based on different demographics. Heart disease is an important example.**

While this dataset is wonderful for a general overview of heart disease, differences based on sex may stem from a bias in the dataset's samples. For example, women represent roughly 1/3 of samples, yet seem to be 300% more likely to NOT have heart disease. Men on the other hand represent roughly 2/3 of samples, but are 122% more likely to have heart disease than not to have it. Multiple resources say cardiovascular disease is the number one killer of both, so we would have an expectation for the amounts to be a bit more even.

### **But how is heart disease declared in this dataset? The original UCI dataset bases the target heart disease on "angiographic disease status"** The values listed on the UCI site say: 

> 58 num: diagnosis of heart disease (angiographic disease status)
-- Value 0: < 50% diameter narrowing
-- Value 1: > 50% diameter narrowing

This can be measured through: "Angiography, angiogram, or arteriograms are terms that describe a procedure used to identify narrowing or blockages in the arteries in the body."
https://www.johnmuirhealth.com/services/cardiovascular-services/intervention/angiography-angiogram-arteriogram.html

### **While this study is older (although the dataset itself is over 2 decades older--ha!), it shows that historically, "... it was found that women are less likely to be referred for functional testing for ischaemia and that a lower rate of diagnostic angiograms and interventional procedures are performed compared with men."**
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3018605/#B2

Which makes me wonder for this dataset--if all of the samples were taken based on the sample population receiving angiographic testing, it would make sense that the women referred to receive this testing (and therefore be part of the sample) could be showing stronger indication of heart disease. Would love to see more background info on this data set and how gathered.

---

Sex is not the only important demographic to base likelihood of heart disease on either. For example, "Non-Hispanic black persons were
more than twice
as likely as non-Hispanic Asian or
Pacific Islander persons to die of
heart disease in 1999 and 2017." https://www.cdc.gov/nchs/hus/spotlight/HeartDiseaseSpotlight_2019_0404.pdf 

### **In future studies and dataset samples on health issues, race should be included as a feature, and there should be a fair representation in sample sizes to help show insight too.**

In addition to future study suggestions, it would be helpful for a larger female sample size. 

Also, smoking is accepted as a major risk factor for heart disease in everyone (although seems to affect women more strongly and in different age groups than men: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3018605/#B2). Smoking was a feature in the original UCI dataset but left out in the main Kaggle dataset. It would be interesting and likely a significant feature to include in future data exploration and modeling.

---

# **So now what? Here are helpful action steps to lower our risk of developing heart disease:**

Based on this dataset, features we may be able to influence include: cholesterol levels (chol), blood pressure (trestbps), and blood sugar (fbs). We can take steps to reach healthy ranges in these areas, whether through medicine prescribed by medical professionals or with lifestyle habits such as: eat plants, stop smoking, get regular exercise and sleep, reduce stress, and get health screenings. https://www.mayoclinic.org/diseases-conditions/heart-disease/in-depth/heart-disease-prevention/art-20046502

**Based on sex differences in the data samples, females should especially work to be their own advocate and seek out preventative testing if they feel they have risk factors.** For example, the high cholesterol outliers were all found in the female group, with ages ranging from 56-67. 

This matches general knowledge of menopause causing cholesterol to rise:

> "Menopause is associated with a progressive increase in total cholesterol, with, in particular, an increase in low-density lipoprotein (LDL), lipoprotein-α and triglycerides and a decrease in high-density lipoprotein (HDL)... Total cholesterol levels peak in women at 55–65 years of age..."  https://www.uscjournal.com/articles/Menopause-cholesterol-cardiovascular-disease

Also: **Heart disease in women less likely to affect large coronary arteries** so general testing often does not pick up on disease in the same way as in men.
> "An interesting observation is that women with ACS have less extensive obstructive and more diffuse coronary artery disease compared with men, but the event rate in nonobstructive coronary artery disease seems to be higher in women"
   https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3018605/

---

**Action steps for researchers include:** plenty of communication about methods, clarity of measures/descriptions for data, and background information can greatly boost understanding.

It also would be worthwhile to compare the original data set from UCI with the data set assigned on Kaggle. Although the data initially looks clean and labeled well, it sounds as though some features may have been assigned to incorrect values over time, so it may be worthwhile to omit those samples if different from the original UCI information.


